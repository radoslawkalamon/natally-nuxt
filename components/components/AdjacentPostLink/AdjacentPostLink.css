.adjacent-post-link {
  color: inherit;
  display: flex;
  flex-flow: column nowrap;
  gap: var(--gap);
  max-width: var(--max-width-adjacent-post-link);
  text-decoration: none;
  width: 100%;
}

.adjacent-post-link__label {
  font-size: var(--font-size-footnotes);
  line-height: var(--line-height-footnotes);
}

.adjacent-post-link__label::after,
.adjacent-post-link__label::before {
  color: var(--color-accent);
}

.adjacent-post-link__title {
  font-family: var(--font-family-sans);
  font-size: var(--font-size-header-3);
  font-weight: bold;
  line-height: var(--line-height-header);
  text-decoration: underline;
}

.adjacent-post-link:hover .adjacent-post-link__title {
  text-decoration: none;
}

.adjacent-post-link--next {
  text-align: right;
}

.adjacent-post-link--next .adjacent-post-link__label::after {
  content: " " var(--content-arrows-right);
}

.adjacent-post-link--previous {
  text-align: left;
}

.adjacent-post-link--previous .adjacent-post-link__label::before {
  content: var(--content-arrows-left) " ";
}
