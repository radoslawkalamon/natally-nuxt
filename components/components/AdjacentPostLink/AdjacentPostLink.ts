import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentsAdjacentPostLink',
  props: {
    label: {
      type: String,
      required: true
    },
    path: {
      type: String,
      required: true
    },
    title: {
      type: String,
      required: true
    },
    type: {
      type: String,
      required: true,
      validator: v => ['next', 'previous'].includes(v)
    }
  },
  computed: {
    classNames () {
      return [
        'adjacent-post-link',
        `adjacent-post-link--${this.type}`
      ]
    }
  }
})
