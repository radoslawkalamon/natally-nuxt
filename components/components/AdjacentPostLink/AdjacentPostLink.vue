<template>
  <NuxtLink
    :class="classNames"
    :to="path"
  >
    <span
      class="adjacent-post-link__title"
      v-text="title"
    />
    <span
      class="adjacent-post-link__label"
      v-text="label"
    />
  </NuxtLink>
</template>

<script src="./AdjacentPostLink.ts" lang="ts"></script>
<style src="./AdjacentPostLink.css" scoped></style>
