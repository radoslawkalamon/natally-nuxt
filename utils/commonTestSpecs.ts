import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
