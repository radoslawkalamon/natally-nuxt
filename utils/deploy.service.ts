/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
