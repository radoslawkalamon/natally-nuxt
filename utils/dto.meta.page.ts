export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
