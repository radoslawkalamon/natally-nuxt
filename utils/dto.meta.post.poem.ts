import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
