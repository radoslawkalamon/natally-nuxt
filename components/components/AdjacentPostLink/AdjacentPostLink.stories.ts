import type { Meta, Story } from '@storybook/vue'
import AdjacentPostLink from '@/components/components/AdjacentPostLink/AdjacentPostLink.vue'

const meta: Meta = {
  title: 'Components / Adjacent Post Link',
  component: AdjacentPostLink,
  argTypes: {
    type: {
      options: ['next', 'previous'],
      control: {
        type: 'inline-radio'
      }
    }
  }
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: { AdjacentPostLink },
  props: Object.keys(argTypes),
  template: '<AdjacentPostLink v-bind="$props" />'
})

export const Next: Story = Template.bind({})
Next.args = {
  label: 'NastÄ™pny',
  path: '/post/test-post/',
  title: 'Post testowy',
  type: 'next'
}

export const Previous: Story = Template.bind({})
Next.args = {
  label: 'Poprzedni',
  path: '/post/test-post/',
  title: 'Post testowy',
  type: 'previous'
}
