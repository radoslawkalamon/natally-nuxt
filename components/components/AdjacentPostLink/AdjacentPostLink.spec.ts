import AdjacentPostLink from '@/components/components/AdjacentPostLink/AdjacentPostLink.vue'
import { shallPassUnitSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory } from '@/devtools/jest.common.spec.utils'

const defaultOptionsFactory = createDefaultOptionsFactory({
  propsData: {
    path: '/post/test-path/',
    title: 'Post testowy'
  }
})

describe('Components / Adjacent Post Link', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: AdjacentPostLink,
      description: 'Next',
      options: defaultOptionsFactory({
        propsData: {
          label: 'NastÄ™pny',
          type: 'next'
        }
      })
    })

    shallPassUnitSanityTest({
      component: AdjacentPostLink,
      description: 'Previous',
      options: defaultOptionsFactory({
        propsData: {
          label: 'Poprzedni',
          type: 'previous'
        }
      })
    })
  })
})
