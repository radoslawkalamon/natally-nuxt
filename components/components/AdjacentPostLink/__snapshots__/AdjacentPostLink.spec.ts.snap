// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Adjacent Post Link Unit shall pass unit sanity test | Next 1`] = `"<nuxtlink-stub to=\\"/post/test-path/\\" class=\\"adjacent-post-link adjacent-post-link--next\\"><span class=\\"adjacent-post-link__title\\">Post testowy</span> <span class=\\"adjacent-post-link__label\\">NastÄ™pny</span></nuxtlink-stub>"`;

exports[`Components / Adjacent Post Link Unit shall pass unit sanity test | Previous 1`] = `"<nuxtlink-stub to=\\"/post/test-path/\\" class=\\"adjacent-post-link adjacent-post-link--previous\\"><span class=\\"adjacent-post-link__title\\">Post testowy</span> <span class=\\"adjacent-post-link__label\\">Poprzedni</span></nuxtlink-stub>"`;
